extensions/data-transfer/portability-data-transfer-google/src/main/java/org/datatransferproject/datatransfer/google/photos/GooglePhotosImporter.java

```diff
@@ -40,92 +42,126 @@
+  static final Logger LOGGER = LoggerFactory.getLogger(GooglePhotosImporter.class);
+
+  private static final String TEMP_PHOTOS_KEY = "tempPhotosData";
 
   private final GoogleCredentialFactory credentialFactory;
   private final JobStore jobStore;
+  private final JsonFactory jsonFactory;
   private final ImageStreamProvider imageStreamProvider;
-  private volatile PicasawebService photosService;
+  private volatile GooglePhotosInterface photosInterface;
 
-  public GooglePhotosImporter(GoogleCredentialFactory credentialFactory, JobStore jobStore) {
-    this(credentialFactory, jobStore, null, new ImageStreamProvider());
+  public GooglePhotosImporter(GoogleCredentialFactory credentialFactory, JobStore jobStore,
+      JsonFactory jsonFactory) {
+    this(credentialFactory, jobStore, jsonFactory, null, new ImageStreamProvider());
   }
 
   @VisibleForTesting
   GooglePhotosImporter(
       GoogleCredentialFactory credentialFactory,
       JobStore jobStore,
-      PicasawebService photosService,
+      JsonFactory jsonFactory,
+      GooglePhotosInterface photosInterface,
       ImageStreamProvider imageStreamProvider) {
     this.credentialFactory = credentialFactory;
     this.jobStore = jobStore;
-    this.photosService = photosService;
+    this.jsonFactory = jsonFactory;
+    this.photosInterface = photosInterface;
     this.imageStreamProvider = imageStreamProvider;
   }
 
   @Override
   public ImportResult importItem(UUID jobId, TokensAndUrlAuthData authData,
-      PhotosContainerResource data) throws IOException, ServiceException {
+      PhotosContainerResource data) throws IOException {
+    // Uploads album metadata
     if (data.getAlbums() != null && data.getAlbums().size() > 0) {
-      logger.warn(
-          "Importing albums in Google Photos is not supported. "
-              + "Photos will be added to the default album.");
+      for (PhotoAlbum album : data.getAlbums()) {
+        importSingleAlbum(jobId, authData, album);
+      }
     }
 
+    // Uploads photos
     if (data.getPhotos() != null && data.getPhotos().size() > 0) {
       for (PhotoModel photo : data.getPhotos()) {
-        importSinglePhoto(authData, photo, jobId);
+        importSinglePhoto(jobId, authData, photo);
       }
     }
 
     return ImportResult.OK;
   }
 
   @VisibleForTesting
-  void importSinglePhoto(TokensAndUrlAuthData authData, PhotoModel inputPhoto, UUID jobId)
-      throws IOException, ServiceException {
-
-    // Set up photo
-    PhotoEntry outputPhoto = new PhotoEntry();
-    outputPhoto.setTitle(new PlainTextConstruct("copy of " + inputPhoto.getTitle()));
-    outputPhoto.setDescription(new PlainTextConstruct(inputPhoto.getDescription()));
-    outputPhoto.setClient(GoogleStaticObjects.APP_NAME);
-
-    String mediaType = inputPhoto.getMediaType();
-    if (mediaType == null) {
-      mediaType = "image/jpeg";
+  void importSingleAlbum(UUID jobId, TokensAndUrlAuthData authData, PhotoAlbum inputAlbum)
```

Comment:
can these methods be private?

Reply:
They have to be package-private since they're being used by tests.

Reply:
oh ok, mind annotating @VisibleForTesting?

Reply:
Looks like it is annotated.

Reply:
sorry! travel brain.. must have read something wrong